Revamped Plugin Reporter due to Librato limitations

ankitcha opened the pull request:
Since librato supports graphs (& alerts) on composite metrics with a maximum limit of 256 metrics - we cannot support recording metrics as partition level - hence changed the design of plugin reporter to record (in cache) & report metrics at group-topic level by aggregating values from all partitions per topic per consumer group.

Also, the earlier version had a bug where gauges were not getting updated value from cache - fixed that as part of new design.

Commit:
Revamped Plugin Reporter due to Librato limitations

Since librato supports graphs (& alerts) on composite metrics with a maximum limit of 256 metrics - we cannot support recording metrics as partition level - hence changed the design of plugin reporter to record (in cache) & report metrics at group-topic level by aggregating values from all partitions per topic per consumer group.

Also, the earlier version had a bug where gauges were not getting updated value from cache - fixed that as part of new design.

dkjhanitt commented on src/main/scala/mix/kafka/monitor/LibratoOffsetReporter.scala:
```diff
@@ -1,86 +1,95 @@
+    val metricsWithUpdatedValues = offsetsToMetrics(offsets)
+    val existingMetrics = cacheProvider.existingCacheEntries(metricsWithUpdatedValues.keys.toSeq)
+    val newMetrics = metricsWithUpdatedValues -- existingMetrics.keys
+    val mergedMetrics = mergeMetrics(existingMetrics, metricsWithUpdatedValues)
```
In general please include `types`. It helps in reviewing faster. Right now I have to scroll up & down to know the exact type.

ankitcha replied:
let me say what I can do. The problem here is - dealing with Map of Map of java types. Including it everywhere makes it even more non-readable. I had it initially and removed it later but tried out following things to solve this concern.

I have created intermediate methods with clear names and explicit return types to try to make it manageable. Other minor thing I intentionally did was to keep the types same. So for instance, in the above method call - `mergeMetrics` both inputs and the return type have same types. Maybe I could include explicit type for the first one to make it better.

dkjhanitt replied:
Maybe `package object` may help here ? Not sure if this is the right use case.

ankitcha replied:
`package object` - ???

ankitcha replied:
instead of using package object - I have type aliases internal within this class

dkjhanitt commented on src/main/scala/mix/kafka/monitor/MetricsCacheProvider.scala:
```diff
@@ -0,0 +1,57 @@
+  logger.info(s"starting MetricsCacheProvider with cacheExpiration = $cacheExpiration")
+
+  def existingCacheEntries(keys: Seq[GroupTopic]): Map[GroupTopic, Map[JInt, JLong]] = {
+    import scala.collection.JavaConverters._
```
Since ` import scala.collection.JavaConverters._` is used in multiple methods, can we move it up ?

ashok12 commented on src/main/scala/mix/kafka/monitor/MetricsCacheProvider.scala:
```diff
@@ -0,0 +1,57 @@
+import kafka.utils.Logging
+import mix.kafka.monitor.LibratoOffsetReporter.getMetricName
+
+case class GroupTopic(group: String, topic: String)
```
Since `GroupTopic` is used across classes, either move it to a separate file or add it to a package object.

Commit:
refactorings as per code review

1. using type aliases for multi-maps
2. using case classes for partition & lag instead of working with java ints & longs
3. minor code optimizations; refactored cache provider to be generic & independent of metrics

ashok12 commented on src/main/scala/mix/kafka/monitor/models.scala:
```diff
@@ -0,0 +1,7 @@
+package mix.kafka.monitor
```
👍
